pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/minion/generator/SegmentMergeRollupTaskGenerator.java

```diff
@@ -0,0 +1,196 @@
+              DEFAULT_MAX_NUM_RECORDS_PER_SEGMENT);
+
+      String bufferTimePeriod =
+          taskConfigs.getOrDefault(MinionConstants.MergeRollupTask.BUFFER_TIME_PERIOD_KEY, DEFAULT_BUFFER_TIME_PERIOD);
+      long bufferTimePeriodMs;
+      try {
+        bufferTimePeriodMs= TimeUtils.convertPeriodToMillis(bufferTimePeriod);
+      } catch (IllegalArgumentException e) {
+        LOGGER.error("Buffer time period ('{}') for table '{}' is not configured correctly.", bufferTimePeriod,
+            offlineTableName, e);
+        throw e;
+      }
+
+      List<OfflineSegmentZKMetadata> segmentsForOfflineTable =
+          _clusterInfoAccessor.getOfflineSegmentsMetadata(offlineTableName);
+
+      // Fetch the segment lineage for the table and compute the segments that should not be scheduled for merge
+      // based on the segment lineage.
+      SegmentLineage segmentLineageForTable = _clusterInfoAccessor.getSegmentLineage(offlineTableName);
+      Set<String> segmentsNotToMerge = new HashSet<>();
+      if (segmentLineageForTable != null) {
+        for (String segmentLineageEntryId : segmentLineageForTable.getLineageEntryIds()) {
+          LineageEntry lineageEntry = segmentLineageForTable.getLineageEntry(segmentLineageEntryId);
+          // Segments shows up on "segmentFrom" field in the lineage entry should not be scheduled again.
+          segmentsNotToMerge.addAll(lineageEntry.getSegmentsFrom());
+
+          // Segments shows up on "segmentsTo" field in the lineage entry with "IN_PROGRESS" state cannot be merged.
+          if (lineageEntry.getState() == LineageEntryState.IN_PROGRESS) {
+            segmentsNotToMerge.addAll(lineageEntry.getSegmentsTo());
+          }
+        }
+      }
+
+      // Filter out the segments that cannot be merged
+      List<SegmentZKMetadata> segmentsToMergeForTable = new ArrayList<>();
+      Set<String> scheduledSegments = scheduledSegmentsMap.getOrDefault(offlineTableName, Collections.emptySet());
+      for (OfflineSegmentZKMetadata offlineSegmentZKMetadata : segmentsForOfflineTable) {
+        String segmentName = offlineSegmentZKMetadata.getSegmentName();
+
+        // The segment should not be merged if it's already scheduled or in progress
+        if (scheduledSegments.contains(segmentName) || segmentsNotToMerge.contains(segmentName)) {
+          continue;
+        }
+
+        // The segments that are newer than the buffer time period should not be be merged
+        if (System.currentTimeMillis() - offlineSegmentZKMetadata.getEndTime() < bufferTimePeriodMs) {
+          continue;
+        }
+        segmentsToMergeForTable.add(offlineSegmentZKMetadata);
+      }
+
+      // Compute Merge Strategy
+      List<List<SegmentZKMetadata>> segmentsToSchedule = MergeStrategyFactory.getMergeStrategy(taskConfigs)
+          .generateMergeTaskCandidates(segmentsToMergeForTable, maxNumSegmentsPerTask);
```

Comment:
Should we also add a parameter for `maxNumTasksPerTable` to limit the number of tasks returned by the merge strategy?